### ISSUE TITLE
No gradients through tfq.math.inner_product()

### ISSUE THREAD

**Issue by c-m-harvey** [tensorflow/quantum#688]
Hi, there are gradient errors when I want to train with keras using  `tfq.math.inner_product()`. I think you can only define gradients for `tfq.layers` unlike in `tf.math` where there are no such gradient issues.  Is there a fix for this? Thanks

**Comment by lockwo** [comment]
I have used inner_product and fidelity in gradient operations before, so I imagine something else is going on in your code. As an example, here is a simple code that takes the gradient through inner_product (and uses other tf ops) and successfully optimizes the fidelity:

```python
import tensorflow as tf
import tensorflow_quantum as tfq
import cirq
import sympy
import numpy as np

symbols = sympy.symbols('alpha beta')
qubits = cirq.GridQubit.rect(1, 2)
reference_circuits = [
    cirq.Circuit(
        cirq.X(qubits[0]) ** symbols[0],
        cirq.Y(qubits[1]) ** symbols[1])
]
other_circuits = [
      cirq.Circuit(
        cirq.X(qubits[0]) ** 0.1,
        cirq.Y(qubits[1]) ** 0.2)
]
reference_tensor = tfq.convert_to_tensor(reference_circuits)
symbol_tensor = tf.convert_to_tensor([s.name for s in symbols])
other_tensor = tfq.convert_to_tensor([other_circuits])
values_tensor = tf.Variable(initial_value=np.random.uniform(0, np.pi, (1, 2)), trainable=True, dtype=tf.float32)

opt = tf.keras.optimizers.Adam(learning_rate=0.1)

for i in range(100):
  with tf.GradientTape() as tape:
    tape.watch(values_tensor)
    ip = tfq.math.inner_product(reference_tensor, symbol_tensor,
                                values_tensor, other_tensor)
    fidelity_loss = 1 - tf.abs(tf.matmul(ip, tf.transpose(ip, conjugate=True)))
  if i % 5 == 0:
    print(i, fidelity_loss.numpy()[0][0])
  grads = tape.gradient(fidelity_loss, values_tensor)
  grads = tf.cast(grads, dtype=tf.float32)
  opt.apply_gradients(zip([grads], [values_tensor]))

print("End fidelity:", tfq.math.fidelity(reference_tensor, symbol_tensor,values_tensor, other_tensor).numpy()[0][0])
```

Output:
```
0 0.7858254
5 0.19569147
10 0.028990328
15 0.07679117
20 0.02313906
25 0.007982373
30 0.02127099
35 0.000970006
40 0.0048906207
45 0.0026465654
50 0.0010214448
55 0.0012292266
60 5.221367e-05
65 0.00056153536
70 7.426739e-05
75 0.00020980835
80 1.9669533e-05
85 7.05719e-05
90 5.9604645e-07
95 2.2292137e-05
End fidelity: 1.0
```

**Comment by c-m-harvey** [comment]
Thanks I see you can achieve this with `tf.GradientTape()`. However, if I just use `tfq.inner_product` as as layer in a keras model, say analogously instead of `keras.layers.Dot()`, it says the gradients don't exist (the model also works with any `tfq.layers` in place of the inner product).

**Comment by lockwo** [comment]
Well everything you can do with model abstractions you can do with GradientTape (or at least, most things I think). That being said, you can also use tfq.inner_product in layers. Here is an example with no gradient tape, that uses classical layers and TFQ layers and successfully minimizes (it is a weird model, but it is just to demonstrate that the gradients can successfully flow through inner product and other classical and quantum layers). You won't be able to just substitute `keras.layers.Dot()` due to the input differences, but it should be possible to use IP in layer/functional models. I think it might just be a problem in your code. 

```python
import tensorflow as tf
import tensorflow_quantum as tfq
import cirq
import sympy
import numpy as np

class test(tf.keras.layers.Layer):
  def __init__(self):
      super(test, self).__init__()
      
      symbols = sympy.symbols('alpha beta')
      qubits = cirq.GridQubit.rect(1, 2)
      symbols1 = [sympy.symbols('gamma')]

      c1 = [cirq.Circuit(cirq.X(qubits[0]) ** symbols[0], cirq.Y(qubits[1]) ** symbols[1])]
      c2 = cirq.Circuit(cirq.X(qubits[0]) ** symbols1[0], cirq.Y(qubits[1]) ** symbols1[0])
      other_circuits = [cirq.Circuit(cirq.X(qubits[0]) ** 0.1, cirq.Y(qubits[1]) ** 0.2)]
      
      self.reference_tensor = tfq.convert_to_tensor(c1)
      self.symbol_tensor = tf.convert_to_tensor([s.name for s in symbols])
      self.other_tensor = tfq.convert_to_tensor([other_circuits])
      self.classical_layer = tf.keras.layers.Dense(2, activation='tanh')
      self.quantum_layer = tfq.layers.ControlledPQC(c2, [cirq.Z(qubits[0]) * cirq.Z(qubits[1])], differentiator=tfq.differentiators.Adjoint())

  def call(self, inputs):
      inputs_classical = inputs[0]
      inputs_quantum = inputs[1]
      classical = self.classical_layer(inputs_classical)
      classical = tf.cast(classical, dtype=tf.complex64)
      ip = tfq.math.inner_product(self.reference_tensor, self.symbol_tensor, classical, self.other_tensor)
      fidelity_loss = tf.abs(tf.matmul(ip, tf.transpose(ip, conjugate=True)))
      return self.quantum_layer([inputs_quantum, fidelity_loss])

opt = tf.keras.optimizers.Adam(learning_rate=0.01)
inputs_classical = tf.keras.layers.Input(shape=(1,))
inputs_quantum = tf.keras.Input(shape=(), dtype=tf.dtypes.string)
mid = test()([inputs_classical, inputs_quantum])
model = tf.keras.models.Model(inputs=[inputs_classical, inputs_quantum], outputs=mid)
model.compile(loss='mse', optimizer=opt)

data = [np.array([[0.2354]]), tfq.convert_to_tensor([cirq.Circuit()])]
label = np.array([[0.0]])

model.fit(x=data, y=label, epochs=10)
```
Output:
```
Epoch 1/10
1/1 [==============================] - 1s 652ms/step - loss: 0.1209
Epoch 2/10
1/1 [==============================] - 0s 10ms/step - loss: 0.0879
Epoch 3/10
1/1 [==============================] - 0s 8ms/step - loss: 0.0609
Epoch 4/10
1/1 [==============================] - 0s 9ms/step - loss: 0.0400
Epoch 5/10
1/1 [==============================] - 0s 10ms/step - loss: 0.0249
Epoch 6/10
1/1 [==============================] - 0s 10ms/step - loss: 0.0146
Epoch 7/10
1/1 [==============================] - 0s 11ms/step - loss: 0.0081
Epoch 8/10
1/1 [==============================] - 0s 6ms/step - loss: 0.0042
Epoch 9/10
1/1 [==============================] - 0s 10ms/step - loss: 0.0021
Epoch 10/10
1/1 [==============================] - 0s 9ms/step - loss: 9.2105e-04
<keras.callbacks.History at 0x7f98edfdbf50>
```

**Comment by c-m-harvey** [comment]
Ok, thanks for the help!
